src/java/org/apache/cassandra/concurrent/DebuggableScheduledThreadPoolExecutor.java

```diff
@@ -41,43 +41,42 @@
 {
     private static final Logger logger = LoggerFactory.getLogger(DebuggableScheduledThreadPoolExecutor.class);
 
-    public static final RejectedExecutionHandler rejectedExecutionHandler = new RejectedExecutionHandler()
-    {
-        public void rejectedExecution(Runnable task, ThreadPoolExecutor executor)
+    public static final RejectedExecutionHandler rejectedExecutionHandler = (task, executor) -> {
+        if (executor.isShutdown())
         {
-            if (executor.isShutdown())
-            {
-                if (!StorageService.instance.isShutdown())
-                    throw new RejectedExecutionException("ScheduledThreadPoolExecutor has shut down.");
+            if (!StorageService.instance.isShutdown())
+                throw new RejectedExecutionException("ScheduledThreadPoolExecutor has shut down.");
 
-                //Give some notification to the caller the task isn't going to run
-                if (task instanceof Future)
-                    ((Future) task).cancel(false);
+            //Give some notification to the caller the task isn't going to run
+            if (task instanceof Future)
+                ((Future) task).cancel(false);
 
-                logger.debug("ScheduledThreadPoolExecutor has shut down as part of C* shutdown");
-            }
-            else
-            {
-                throw new AssertionError("Unknown rejection of ScheduledThreadPoolExecutor task");
-            }
+            logger.debug("ScheduledThreadPoolExecutor has shut down as part of C* shutdown");
+        }
+        else
+        {
+            throw new AssertionError("Unknown rejection of ScheduledThreadPoolExecutor task");
         }
     };
 
     public DebuggableScheduledThreadPoolExecutor(int corePoolSize, String threadPoolName, int priority)
     {
         super(corePoolSize, new NamedThreadFactory(threadPoolName, priority));
+        setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
         setRejectedExecutionHandler(rejectedExecutionHandler);
     }
 
     public DebuggableScheduledThreadPoolExecutor(int corePoolSize, ThreadFactory threadFactory)
     {
         super(corePoolSize, threadFactory);
+        setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
         setRejectedExecutionHandler(rejectedExecutionHandler);
     }
 
     public DebuggableScheduledThreadPoolExecutor(String threadPoolName)
     {
         this(1, threadPoolName, Thread.NORM_PRIORITY);
+        setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
```

Comment:
I am just looking at the same as I saw CI failures which I am looking at one by one now. I think what you said would be safer but let me dig into it a bit more now

Reply:
> how about just call shutdownNow() in the decommission method? It has the same effect

That's not exactly true. `shutdownNow()` interrupts running tasks.

Reply:
I think `setExecuteExistingDelayedTasksAfterShutdownPolicy()` only affects delayed tasks, and if we're relying on delayed tasks running for correctness, we have much bigger problems to worry about. If you look at places like `SnapshotDeletingTask`, we submit tasks instead of scheduling them, and in the approach where we `shutdown()`, we allow tasks like that to complete.

Reply:
I think I am more and more in for this proposal https://github.com/ekaterinadimitrova2/cassandra/pull/97#discussion_r602603095

Reply:
I did the change here - https://github.com/ekaterinadimitrova2/cassandra/commit/ff230d450f373e2c6ae0389364e31141387d9fc0

CI running here - https://jenkins-cm4.apache.org/job/Cassandra-devbranch/541/